feat: Kafka broker

vividvilla opened the pull request:
Here is a implementation of Kafka broker, I went through other broker implementations and tried to implement something similar but with few changes like how the worker pool is initialized etc. Also introduced some opinionated changes to initialize Kafka broker
- Broker url is not required
- New config struct `config.KafkaConfig` 

Here is an sample initialization

```go
cnf := &config.Config{
  DefaultQueue:    "KAFKA_TOPIC_TO_LISTEN",
  ResultBackend:   "YOUR_BACKEND_URL",
  Kafka: &config.KafkaConfig{
    // Kafka consumer group name.
    Group:    "KAFKA_GROUP",
    // Kafka broker URLs.
    Addrs:    []string{"127.0.0.1:9092"},
    // Client ID used by Kafka client (optional).
    ClientID: "KAFKA_CLIENT_ID",
    // Consume from newest if enabled, bydefault it consumes from oldest.
    OffsetNewest: true,
    // Set compression format (optional). Allowed values are `gzip`, `lz4`, `snappy`, `zstd`.
    Compression: "gzip",
    // Set compression level (optional).
    CompressionLevel: 5,
  },
}
```

This pull request is based on `v1.6.7`.

edit: Added client level compression.

vividvilla committed:
feat: Kafka broker

vividvilla committed:
feat: add kafka client compression

vividvilla committed:
fix: rebase master and fix merge conflict

vividvilla commented:
Update the fork with latest changes to upstream master. Hope this gets reviewed and merged soon :)

vividvilla committed:
feat: add kafka message size config

RichardKnop commented:
@vividvilla I'd like first to solve the problem of importing all dependencies as this is becoming bigger issue. By adding more brokers/backends, it increases amount of dependencies that machinery will import.

It would be best to refactor the code so it only imports dependencies of broker/backend you are using.

Little bit afraid of adding yet another broker before that is addressed.

vividvilla commented:
I am glad that you acknowledged the issue in current system, when I started working with Machinery I was so shocked by the number of dependencies it pulls and how tightly coupled broker configs are. Recently our CI build were started failing because of Apache thrift library fetch timeout , we had to switch to Go mod proxy to solve the issue. Also compile time is very high compared to any of our other Go projects and binary size also 3-4x bloated. 

In fact I started refactoring my Machinery fork to address these issues but halted for a while due to time constraint on our company projects. My idea was to ditch current way of passing broker url to initialize server and instead send broker itself, for example

```go
server, err := machinery.NewServer(&config.Config{
	Broker: kafkabroker.New(kafkabroker.Config{
		Group:    "broker_group",
		Addrs:    ["broker_addrs"],
		ClientID: "broker_id",
	})
})
```

Similarly backend stores also has to be initialized and passed on as config.

Also the idea is to remove support for current Yaml config, ideally user should be using projects like [Viper](https://github.com/spf13/viper/) or [Koanf](https://github.com/spf13/viper/) for configuration, it shouldn't be a part of the library. 

These changes I have proposed are not backward compatible so has to be released as new version.

RichardKnop commented:
@vividvilla Something like that would be ideal. I am still thinking about a ways to refactor the code to make it not import all dependencies (even those you don't need) without breaking changes. But it might have to be a breaking change in which case I would probably create a new top level package called v2.

RichardKnop commented:
Possibly, we could start with a v2 package like this: https://github.com/RichardKnop/machinery/pull/477

Still basically compatible with current version but there is no factory so when you import it from your project, you will only need to import dependencies for a broker and backend you are using.

vividvilla committed:
fix(kafka broker): check if message is null before sending it to task queue

vividvilla committed:
fix(kafka broker): partitioned consumer group doesn't stop consuming

vividvilla committed:
fix: number of worker pool increased by 1

vividvilla committed:
fix(kafka broker): task error channel not being consumed

marcocali commented on v1/brokers/kafka/kafka.go:
```diff
@@ -0,0 +1,298 @@
+	"sync"
+	"sync/atomic"
+
+	"github.com/RichardKnop/machinery/v1/brokers/errs"
```
Please organize imports in 3 different groups:

- Library
- 3rd party
- Local

Same as described here: https://github.com/golang/tools/commit/ed69e84b1518b5857a9f4e01d1f9cefdcc45246e

vividvilla committed:
feat: add cassandra backend support

vividvilla committed:
Merge pull request #1 from vividvilla/cassandra-backend

feat: add cassandra backend support

iamd3vil committed:
fix: Add changes for compatibility with both Redis and SSDB.

vividvilla committed:
Merge pull request #2 from iamd3vil/fix/changes-for-ssdb

fix: Add changes for compatibility with both Redis and SSDB.

vividvilla committed:
feat: goredis backend

vividvilla committed:
Merge pull request #3 from vividvilla/goredis-backend

feat: goredis backend

vividvilla committed:
fix: don't supress the error returned on new backend

yuexian1234 commented:
lgtm

vividvilla committed:
feat(kafka): add TLS support for producer and consumer

Pass TLS config from main config to both Kafka producer and consumer
group, enabling secure connections when TLS configuration is provided.
